server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java

```diff
@@ -61,7 +70,7 @@
   private final CompactionSegmentSearchPolicy policy;
   private final IndexingServiceClient indexingServiceClient;
 
-  private Object2LongOpenHashMap<String> totalSizesOfSegmentsAwaitingCompactionPerDataSource;
+  private HashMap<String, AutoCompactionSnapshot> autoCompactionSnapshotPerDataSource = new HashMap<>();
```

Comment:
Changed AutoCompactionSnapshot to be immutable and changed autoCompactionSnapshotPerDataSource to be a concurrent hash map.

Reply:
When a new compaction run happens, we create a new snapshot and update (atomically) the autoCompactionSnapshotPerDataSource since autoCompactionSnapshotPerDataSource is a concurrent hash map.

Reply:
Changed the autoCompactionSnapshotPerDataSource into a AtomicReference<Map<String, AutoCompactionSnapshot>>.

We will build a new autoCompactionSnapshotPerDataSource of snapshots in each compaction run then replace (atomically) the whole map (autoCompactionSnapshotPerDataSource) at the end in `makeStats`